### PR TITLE
Fix span write in `InvariantCreateSortKeyOrdinalIgnoreCase`

### PR DESCRIPTION
**Pull request by xtqqczze**
cc: @tarekgh

Existing logic appears to be incorrect:

```csharp
BinaryPrimitives.WriteUInt16BigEndian(sortKey, hr);
BinaryPrimitives.WriteUInt16BigEndian(sortKey, lr)
```

### CODE REVIEW

**Comment by tarekgh** [comment]
@xtqqczze Thanks for your contribution. At the moment, we are only accepting critical changes for the .NET 9.0 release. Nice-to-have fixes will need to wait until we fork the main branch to accept changes for future releases. Since this PR is already a draft, we can keep it open for a while until the fork. Alternatively, you can close it and reopen it later after the fork. I wanted to set the expectation in case you notice no traction on the PR.

CC @ericstj @stephentoub

**Comment by xtqqczze** [comment]
@tarekgh This fixes a probable bug.

**Comment by tarekgh** [comment]
@xtqqczze I'll take a look. Thanks!

**Comment by tarekgh** [comment]
@xtqqczze could you please explain more why you think there is a bug? 

`InvariantCreateSortKeyOrdinalIgnoreCase` is called guaranteeing the condition `Debug.Assert(sortKey.Length >= source.Length * sizeof(char));`.

**Comment by xtqqczze** [comment]
> @xtqqczze could you please explain more why you think there is a bug?

@tarekgh When writing to the span, we overwrite the high surrogate with the low surrogate:

https://github.com/dotnet/runtime/blob/f644110e477b942636c6a526d0cdf688c737a069/src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Invariant.cs#L69-L70

This is unexpected, at least.

**Comment by stephentoub** [comment]
Can you share a test that fails because of the issue?

**Comment by tarekgh** [comment]
> When writing to the span, we overwrite the high surrogate with the low surrogate:

I see. 

> Can you share a test that fails because of the issue?

This is a good idea to add a test for this case.

**Comment by tarekgh** [comment]
Here is the code can be used in the test case:

```C#
            CultureInfo culture = CultureInfo.InvariantCulture;
            var sortKey = culture.CompareInfo.GetSortKey("\uD800\uDC00", CompareOptions.IgnoreCase);

            foreach (var b in sortKey.KeyData)
            {
                Console.Write($"{b:X2} "); // should produce something like D8-00-DC-00
            }
````

We need to add the test to https://github.com/dotnet/runtime/tree/main/src/libraries/System.Runtime/tests/System.Globalization.Tests/Invariant

**Comment by xtqqczze** [comment]
This is technically a breaking change to [GetSortKey(ReadOnlySpan<Char>, Span<Byte>, CompareOptions)](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.getsortkey?view=net-8.0#system-globalization-compareinfo-getsortkey(system-readonlyspan((system-char))-system-span((system-byte))-system-globalization-compareoptions)) in the _Bucket 3: Unlikely Grey Area_ category.

**Comment by xtqqczze** [comment]
@MihuBot

**Comment by tarekgh** [comment]
> This is technically a breaking change to [GetSortKey(ReadOnlySpan, Span, CompareOptions)](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.getsortkey?view=net-8.0#system-globalization-compareinfo-getsortkey(system-readonlyspan((system-char))-system-span((system-byte))-system-globalization-compareoptions)) in the Bucket 3: Unlikely Grey Area category.

I am not worried about that. The issue is specific for Invariant Globalization mode with Surrogate pairs. Sort keys usually interesting with the full globalization support.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Invariant.cs [hunk 1]
```diff
@@ -66,10 +66,11 @@ private static void InvariantCreateSortKeyOrdinalIgnoreCase(ReadOnlySpan<char> s
                     if (char.IsLowSurrogate(cl))
                     {
                         SurrogateCasing.ToUpper(c, cl, out char hr, out char lr);
-                        BinaryPrimitives.WriteUInt16BigEndian(sortKey, hr);
-                        BinaryPrimitives.WriteUInt16BigEndian(sortKey, lr);
-                        i++;
+                        Span<byte> tmp = sortKey.Slice(0, 2 * sizeof(ushort)); // help with bounds check elimination
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Invariant.cs:69]: Why is this needed? Why is it not enough to use the SortKey span directly?

Comment (xtqqczze) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Invariant.cs:69]: By slicing the span before writing, we can avoid two bounds checks:

https://www.diffchecker.com/UFgsRYlS/

[godbolt](https://csharp.godbolt.org/z/GvfWsoocs)

**Review by tarekgh — APPROVED** [review]
Thanks @xtqqczze!

**Review by stephentoub — APPROVED** [review]
